tests/CallbacksTest.test.nut

```diff
@@ -0,0 +1,173 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO
+// EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+// OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+@include "./ConnectionManager.lib.nut"
+
+class CallbacksTest extends ImpTestCase {
+
+  cm = null;
+
+  function setUp() {
+    this.info("running setUp");
+    cm = ConnectionManager();
+  }
+
+  /*
+  * sets onDisconnect callback and disconnects device using CM
+  *
+  */
+  function testOnDisconnectAsync() {
+    return Promise(function(resolve, reject) {
+      cm.onDisconnect(function(expected) {
+        cm.connect();
+      }.bindenv(this));
+
+      cm.onConnect(function() {
+        resolve();
+      }.bindenv(this));
+
+      cm.disconnect();
+      assertTrue(!server.isconnected(), "should NOT be connected!");
+    }.bindenv(this))
+    .then(function(val) {
+        assertTrue(server.isconnected(), "should be connected again!");
+        _resetCM.call(this);
+    }.bindenv(this))
+    .fail(function(valueOrReason) {
+        _resetCM.call(this);
+        throw valueOrReason;
+      }.bindenv(this));
+  }
+
+ /*
+  * sets onConnect callback and disconnects device using CM
+  *
+  */
+  function testOnConnectAsync() {
+    local flag = 0;
+    return Promise(function(resolve, reject) {
+      cm.onDisconnect(function(expected) {
+        assertTrue(!flag, "should be 0 now");
+        cm.connect();
+      }.bindenv(this));
+
+      cm.onConnect(function() {
+        flag++;
+        resolve();
+      }.bindenv(this));
+
+      assertTrue(!flag, "should 0 before re-connect");
+      cm.disconnect();
+      assertTrue(!server.isconnected(), "should NOT be connected!");
+    }.bindenv(this))
+    .then(function(val) {
+      //not assertTrue(flag, ...) to ensure flag was increased exactly one time
+      assertTrue(flag==1, "connected again, should be 1 now");
+      _resetCM.call(this);
+    }.bindenv(this))
+    .fail(function(valueOrReason) {
+      _resetCM.call(this);
+      throw valueOrReason;
+    }.bindenv(this));
+  }
+
+ /*
+  * sets onNextConnect callback and disconnects device using CM
+  *
+  */
+  function testOnNextConnectAsync() {
+    local flag = 0;
```

Comment:
It can be a local variable and returned to the .then statement through the resolve(flag) call.
I would also rename it to `step` or something like that